File: typescript/rna-transcription/rna-transcription.ts
const enum Dna {
  C = "C",
  G = "G",
  A = "A",
  T = "T"
}

const enum Rna {
  G = "G",
  C = "C",
  U = "U",
  A = "A"
}

type MaybeError<T> = T | never
type RnaTranscriptions = Readonly<Record<Dna, Rna>>

export default class Transcriptor {
  private readonly RNA_TRANSCRIPTIONS: RnaTranscriptions = {
    [Dna.C]: Rna.G,
    [Dna.G]: Rna.C,
    [Dna.A]: Rna.U,
    [Dna.T]: Rna.A
  }

  toRna(dna: string): string {
    return (
      dna
        .split("")
        .map(this.translateNucleotide.bind(this))
        .join("")
    )
  }

  private translateNucleotide(nucleotide: string): MaybeError<Rna> {
    return this.RNA_TRANSCRIPTIONS[nucleotide as Dna] || this.throwInvalidDna()
  }

  private throwInvalidDna(): never {
    throw new Error("Invalid input DNA.")
  }
}

File: python/rna-transcription/rna_transcription.py
from typing import Dict, List
from enum import Enum


class Dna(Enum):
    C = "C"
    G = "G"
    A = "A"
    T = "T"


class Rna(Enum):
    G = "G"
    C = "C"
    U = "U"
    A = "A"


__RNA_TRANSCRIPTIONS: Dict[Dna, Rna] = {
    Dna.C: Rna.G,
    Dna.G: Rna.C,
    Dna.A: Rna.U,
    Dna.T: Rna.A
}


def to_rna(dna_strand: str) -> str:
    rna_nucleotides: List[str] = list(
        map(__translate_nucleotide, list(dna_strand))
    )
    return "".join(rna_nucleotides)


def __translate_nucleotide(nucleotide: str) -> str:
    return Rna(__RNA_TRANSCRIPTIONS[Dna(nucleotide)]).value
